### PR TITLE
Process concs

### PR DESCRIPTION
**Pull request by znichollscr**
Add basic processing of concs. This runs from nothing to having data files written in input4MIPs format. The actual written data is more or less just a placeholder, but this helps us check workflows and data formats.

I've moved the question about data retrieval to #3

### CODE REVIEW

**Comment by znichollscr** [comment]
Alrighty @lewisjared and @mikapfl, ready for review. The only essential things are:

- read the README, see if it makes sense and the suggested commands run
- run `make dev`, make sure it runs through

If you want to comment on anything else, you've very welcome, but no hurry at this stage

File: tests/regression/test_workflow/test_basic_workflow.npz
Comment (lewisjared) [inline tests/regression/test_workflow/test_basic_workflow.npz:1]: I'm assuming this is the file generated by pytest-regression? I've never seen that format before, but I guess it doesn't matter

File: notebooks/01yy_process-data/0101_process-law-dome.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+pdf["time"] = pdf["time"] / (seconds_per_year) + 1970
+pdf
+
+# %%
+gas = "N2O"
+law_dome = ColumnDataSource(
+    pdf.loc[pdf["variable"].isin([f"Atmospheric Concentrations|{gas}"])]
+)
+
+figure_bokeh = figure(
```
Comment (lewisjared) [inline notebooks/01yy_process-data/0101_process-law-dome.py:130]: Haven't run locally, butdo these figures remain interactive in the archived ipynb?

**Review by lewisjared — APPROVED** [review]
I can't really add anything critical here as there is a lot coming and going. Why strip out all of the 1xx+ notebooks? Were they just placeholders?

**Comment by znichollscr** [comment]
> Why strip out all of the 1xx+ notebooks? Were they just placeholders?

Yep they're just unhelpful noise here. They're good for a copier template, but here they just add runtime for zero benefit.

File: notebooks/00yy_retrieve-data/0001_law-dome.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+# %%
+print(f"DOI for dataset: {config_step.law_dome.doi}")
+
+# %% [markdown]
+# ## Retrieve the data
+#
+# **Note**: We would like to automate this using the command below. However, the keys are rotated.
+#
+# ```
+# !AWS_ACCESS_KEY_ID=ADROIMK8WFURFMTPLD9A AWS_SECRET_ACCESS_KEY=woGxqR3TD5gmn1/ICDEp9G8iQLhm968IqHtV0rF0 aws s3 cp --endpoint-url https://s3.data.csiro.au --recursive s3://dapprd/000037077v001/ {config_step.law_dome.raw_dir}
```
[the remark this answers is not in the document]
Comment (znichollscr) [reply notebooks/00yy_retrieve-data/0001_law-dome.py:62]: Nope. The secret key isn't 'mine' (you just get given it when clicking on CSIRO's download stuff) so this one isn't an issue, but yes would be nice to have better catching of this in general.

[the remark this answers is not in the document]
Comment (znichollscr) [reply notebooks/00yy_retrieve-data/0001_law-dome.py:62]: Will you make an issue somewhere to follow up (I assume not here as this isn't a repo specific problem)?

File: tests/regression/test_workflow/test_basic_workflow.npz
Comment (znichollscr) [reply tests/regression/test_workflow/test_basic_workflow.npz:1]: Yep, numpy's file format https://imageio.readthedocs.io/en/v2.5.0/format_npz.html#npz-numpy-s-compressed-array-format

Comment (znichollscr) [reply tests/regression/test_workflow/test_basic_workflow.npz:1]: I'm assuming it's portable given tests don't fail and pytest-regressions use it...

File: notebooks/01yy_process-data/0101_process-law-dome.py (hunk 1 above)
Comment (znichollscr) [reply notebooks/01yy_process-data/0101_process-law-dome.py:130]: Nope. But you can always just re-run the notebook in place to get the interactivity back.

Can you please try running the whole thing following the README? The runtime is short (like 3 minutes) and I would like feedback on whether the README makes any sense or not and whether this works out of the box or whether I'm missing some key step (full request here https://github.com/climate-resource/CMIP-GHG-Concentration-Generation/pull/2#issuecomment-1852318690).

File: notebooks/00yy_retrieve-data/0001_law-dome.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+#       format_version: '1.3'
+#       jupytext_version: 1.15.2
+#   kernelspec:
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# # Law dome
```
[the remark this answers is not in the document]
Comment (znichollscr) [reply notebooks/00yy_retrieve-data/0001_law-dome.py:16]: Updated (a DOI appears later in the notebook but I've now made it more obvious)

File: notebooks/01yy_process-data/0101_process-law-dome.py (hunk 1 above)
Comment (lewisjared) [reply notebooks/01yy_process-data/0101_process-law-dome.py:130]: Works a treat and readme makes sense

Comment (znichollscr) [reply notebooks/01yy_process-data/0101_process-law-dome.py:130]: Thanks

File: notebooks/00yy_retrieve-data/0001_law-dome.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lewisjared) [reply notebooks/00yy_retrieve-data/0001_law-dome.py:64]: Addded issue to copier. I couldn't find a good soln in 5 minutes of looking

**Review by mikapfl — APPROVED** [review]
Hi,

Both the "normal" and the "dev" workflow worked for me, and produced something which looks sensible.

I looked around the source code a bit and with the context I already had about your notebook setup etc., I found the notebooks quite nice and easy to follow. What I found mildly confusing is to know how much the code in the `pydoit_nb` folder is project-specific or if it just vendored non-specific code. I guess it is all non-specific, but maybe not.

For the README, I think one or two additional sentences about the scope of the project would be great. It currently says:

> Generation of historical concentrations for CMIP (and other) experiments

which nicely specifies the outputs of this, but doesn't tell about the inputs. I guess this Python project doesn't go out itself and collects ice cores. :wink: 
So, maybe on the level of "We use multiple different historical concentration data sets and synthesize a consistent set" or "We use input datasets with global resolution and downscale to finer grids" would be great. Essentially, what is the scientifically most exciting part in one or two sentences.

**Comment by mikapfl** [comment]
Something which the latest ruff complains about and I agree: You are using `subprocess.run` without the `check` argument. It is good practice to be explicit about the fact that you either don't care about errors in the called programs (the default) or want to raise Exceptions when things go wrong.

**Comment by znichollscr** [comment]
Super nice feedback thanks!

> I guess it is all non-specific, but maybe not

My hope is that all of pydoit_nb is non-specific, but I guess I will only find out once I actually try and pull it out (a Q1 job as part of this work).

> For the README, I think one or two additional sentences about the scope of the project would be great

Nice, will add.

> Something which the latest ruff complains about and I agree: You are using subprocess.run without the check argument. It is good practice to be explicit about the fact that you either don't care about errors in the called programs (the default) or want to raise Exceptions when things go wrong.

Nice. I was confused about how to clarify, "just run, stop complaining ruff". Thanks for the pointer (I obviously didn't look very hard as a quick google would have probably shown me the answer, maybe I was googling wrong, I dunno), I will add.

**Comment by znichollscr** [comment]
All added, thanks again both. I'll merge then keep rolling in the new year.

**Pull request merged**